bigflow/dataflow.py

```diff
@@ -1,69 +1,204 @@
+        if not self.entry_point:
+            raise RuntimeError("You need override method 'init_pipeline' or provide 'entry_point' argument")
+        self.entry_point(pipeline, context, *self.entry_point_args, **self.entry_point_kwargs)
 
-    @staticmethod
-    def _apply_logging(pipeline_options: PipelineOptions, workflow_id: str) -> PipelineOptions:
-        google_cloud_options = pipeline_options.view_as(GoogleCloudOptions)
-        if not google_cloud_options.labels:
-            google_cloud_options.labels = []
-        google_cloud_options.labels.append(f'workflow_id={workflow_id}')
-        return pipeline_options
+    def run_pipeline(self, context: JobContext, pipeline: Pipeline):
+        logger.info("Run pipeline, context %s", context)
+        return pipeline.run()
+
+    def create_pipeline_options(self, context: JobContext) -> PipelineOptions:
+        options = dict(self.pipeline_options)
+        self.set_default_pipeline_options(context, options)
+        logger.debug("Pipeline options from dict %s", options)
+        return PipelineOptions.from_dictionary(options)
+
+    def set_default_pipeline_options(self, context: JobContext, options: PipelineOptionsDict):
+        if not self.no_default_pipeline_options:
+            return
+
+        logger.debug("Add defaults to pipeline options")
+
+        if 'job_name' not in options:
+            slug = self.id.replace("_", "-")
+            job_name = f"{slug}-{uuid.uuid4().hex}"
+            logger.info("Set job name to deafult %s", job_name)
+            options['job_name'] = job_name
+
+        if 'setup_file' not in options:
+            setuppy = bigflow.build.reflect.materialize_setuppy(self._project_path)
+            logging.debug("Add setup.py %s to pipeline options", setuppy)
+            options['setup_file'] = setuppy
+        else:
+            logging.debug("Pipeline options already contains 'setup_file': %s", options['setup_file'])
+
+        workflow_id = context.workflow_id
+        if workflow_id:
+            options['labels'] = (options.get('labels') or []) + [f'workflow_id={workflow_id}']
+        else:
+            logger.warning("A workflow_id is not found in the context - skip logging initialization.")
+
+        if self.use_docker_image:
+            if isinstance(self.use_docker_image, str):
+                logger.debug("Use explicitly provided docker image")
+                imgid = self.use_docker_image
+            else:
+                logger.debug("Infer docker image name for current project")
+                pspec = bigflow.build.reflect.get_project_spec(self._project_path)
+                imgid = build_docker_image_tag(pspec.docker_repository, pspec.version)
+            logger.info("Use docker image %s for beam workers", imgid)
 
+            options['worker_harness_container_image'] = imgid
+            experiments = options.setdefault('experiments', [])
+            if 'use_runner_v2' not in experiments:
```

Comment:
I'm sure if that should be a default, it's buggy

Reply:
sure or you meant not sure?

Reply:
Yep, it mutates original 'pipeline_options['experiments']`list. Fixed.